pilot/pkg/config/coredatamodel/controller.go

```diff
@@ -175,17 +176,40 @@ func (c *Controller) Apply(change *sink.Change) error {
 		}
 	}
 
-	var prevStore map[string]map[string]*model.Config
-
 	c.configStoreMu.Lock()
-	prevStore = c.configStore[descriptor.Type]
 	c.configStore[descriptor.Type] = innerStore
 	c.configStoreMu.Unlock()
 
-	if descriptor.Type == model.ServiceEntry.Type {
-		c.serviceEntryEvents(innerStore, prevStore)
+	if descriptor.Type == model.ServiceEntry.Type &&
+		descriptor.Collection == model.ServiceEntry.Collection {
+		for ns, byName := range innerStore {
+			for name, config := range byName {
+				endpoints := []*model.IstioEndpoint{}
+				if se, ok := config.Spec.(*networking.ServiceEntry); ok {
+					for _, ep := range se.Endpoints {
+						for portName, port := range ep.Ports {
+							ep := &model.IstioEndpoint{
+								Address:         ep.Address,
+								EndpointPort:    port,
+								ServicePortName: portName,
+								Labels:          ep.Labels,
+								UID:             fmt.Sprintf("%s.%s", name, ns),
+								// TODO: how to get network and ServiceAccount?
+							}
+							endpoints = append(endpoints, ep)
+						}
+					}
+				}
+				// we know everything associated to this type needs an update
+				// so no need to check version or equality of the endpoints
+				hostname := fmt.Sprintf("%s.%s.svc.%s", name, ns, c.options.DomainSuffix)
+				c.options.XDSUpdater.EDSUpdate("MCP", hostname, endpoints)
```

Comment:
@costinm @nmittler  - do hosts (services) ever need to be removed? This logic only seems to add/append.

Reply:
I think the delete case should be handled when we receive the other type of service entry which causes a full computation of the configs. @costinm @nmittler Is that the case?

Reply:
FYI, this logic is incorrect. You should trigger EDSUpdate if and only if model.Service. lb type is clientsideLB. Service entry's endpoints can have hostnames which are not allowed in EDS. [eds takes only ips].
There is an involved inference logic in the service entry conversion code that determines whether a service entry will be treated as clientSideLB, strictDNS or passthrough. Please use it

Reply:
Also, service entries also trigger config updates if you are adding a new service, or removing a service

Reply:
ConfigUpdate is still triggered depending on the type of serviceEntries (synthetic/non synthetic) we receive which will cause a full computation of the configs.

Reply:
@ayj To answer your question: in the absence of incremental MCP, we agreed the functionality is like this:

- the addition of a ServiceEntry is signaled by send an MCP update with the full list of ServiceEntries without that ServiceEntry
- the deletion of a ServiceEntry is signaled by sending an MCP update with the full list of ServiceEntries without that ServiceEntry

For Endpoints, the MCP update will contain a single ServiceEntry and the full list of Endpoints.

Reply:
Incremental MCP has no impact on this. 

Pilot expects a full list of endpoints associated with one Service, or an empty list if the Service has no more endpoints.

If the service itself is deleted - a full config computation is done, and should cleanup. 

Shriram is right - but I don't think it's relevant for this PR, which takes the code we currently use in 
pilot for endpoints, and generates an identical behavior using MCP. I agree in future we'll need to 
improve - any ServiceEntry that is not generated from Endpoints should stay in the normal config code path. 

In other words for the new type we need a 1-1 replacement of the k8s registry, with identical behavior. We can improve or fix other bugs - but let's not mix things. 

Incremental EDS is intended for only one thing - when the list of endpoints associated with a service in k8s or other discovery system has changed. Since this is a frequent event - far more frequent than user editing configs, and with a larger set of data - it has special optimization in pilot and envoy. Has nothing to do with incremental MCP - it doesn't matter if MCP sends the endpoints incrementally or not ( and I can't imagine the average 3-4 entries that make up a service sent efficiently as incremental instead of a single proto). Incremental EDS only means that 
envoy sends the load assignment for a cluster whenever it needs to, without having to send other clusters that didn't change.